src/bespokelabs/curator/request_processor/base_online_request_processor.py

```diff
@@ -0,0 +1,467 @@
+                        continue
+
+                    request = APIRequest(
+                        task_id=status_tracker.num_tasks_started,
+                        generic_request=generic_request,
+                        api_specific_request=self.create_api_specific_request(generic_request),
+                        attempts_left=max_attempts,
+                        prompt_formatter=self.prompt_formatter,
+                    )
+
+                    token_estimate = self.estimate_total_tokens(request.generic_request.messages)
+
+                    # Wait for capacity if needed
+                    while not status_tracker.has_capacity(token_estimate):
+                        await asyncio.sleep(0.1)
+
+                    # Consume capacity before making request
+                    status_tracker.consume_capacity(token_estimate)
+
+                    task = asyncio.create_task(
+                        self.handle_single_request_with_retries(
+                            request=request,
+                            session=session,
+                            retry_queue=queue_of_requests_to_retry,
+                            save_filepath=save_filepath,
+                            status_tracker=status_tracker,
+                        )
+                    )
+                    pending_requests.append(task)
+
+                    status_tracker.num_tasks_started += 1
+                    status_tracker.num_tasks_in_progress += 1
+
+            # Wait for all tasks to complete
+            if pending_requests:
+                await asyncio.gather(*pending_requests)
+
+            # Process any remaining retries in the queue
+            while not queue_of_requests_to_retry.empty():
+                retry_request = await queue_of_requests_to_retry.get()
+                token_estimate = self.estimate_total_tokens(retry_request.generic_request.messages)
+
+                attempt_number = 6 - retry_request.attempts_left
+                logger.info(
+                    f"Processing final retry for request {retry_request.task_id} "
+                    f"(attempt #{attempt_number} of 5). "
+                    f"Previous errors: {retry_request.result}"
+                )
+
+                # Wait for capacity if needed
+                while not status_tracker.has_capacity(token_estimate):
+                    await asyncio.sleep(0.1)
+
+                # Consume capacity before making request
+                status_tracker.consume_capacity(token_estimate)
+
+                task = asyncio.create_task(
+                    self.handle_single_request_with_retries(
+                        request=retry_request,
+                        session=session,
+                        retry_queue=queue_of_requests_to_retry,
+                        save_filepath=save_filepath,
+                        status_tracker=status_tracker,
+                    )
+                )
+                await task
```

Comment:
won't this await blocking the entire loop here until the task is done?

i think what we want is actually to do await asyncio.sleep (https://github.com/bespokelabsai/curator/blob/112cd91895a24a03dfe070a79d0dc0722da73142/src/bespokelabs/curator/request_processor/openai_online_request_processor.py#L417C1-L417C21) so that it can switch to other tasks. otherwise this will only submit one task at a time.

could you run @RyanMarten's speed test to make sure performance doesn't drop?

Reply:
I ran the test, and didn't see any issues

Reply:
Performance drop happens when processing the retried entries. I compared two versions, and the current implementation is noticably slower. Only noticable when there's 10+ retried entries. I'll make a fix.